core/src/main/java/com/linecorp/armeria/common/DefaultHttpResponseBodyStream.java

```diff
@@ -0,0 +1,352 @@
+                } else if (completing) {
+                    downstream.onComplete();
+                }
+            } catch (Throwable t) {
+                throwIfFatal(t);
+                logger.warn("Subscriber should not throw an exception. subscriber: {}", downstream, t);
+            }
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (!upstreamUpdater.compareAndSet(this, null, subscription) || cancelCalled) {
+                subscription.cancel();
+                return;
+            }
+            if (pendingRequests != 0) {
+                subscription.request(pendingRequests);
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                // Just abort the publisher so subscriber().onError(e) is called and resources are cleaned up.
+                response.abort(new IllegalArgumentException(
+                        "n: " + n + " (expected: > 0, see Reactive Streams specification rule 3.9)"));
+                return;
+            }
+            final Subscription upstream = this.upstream;
+            if (upstream == null) {
+                pendingRequests = LongMath.saturatedAdd(n, pendingRequests);
+            } else {
+                upstream.request(n);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            if (cancelCalled) {
+                return;
+            }
+            cancelCalled = true;
+            downstream = NoopSubscriber.get();
+            maybeCompleteHeaders();
+            if (upstream != null) {
+                upstream.cancel();
+            }
+        }
+
+        @Override
+        public void onNext(HttpObject httpObject) {
+            final Subscription upstream = this.upstream;
+            if (httpObject instanceof ResponseHeaders) {
+                final ResponseHeaders headers = (ResponseHeaders) httpObject;
+                final HttpStatus status = headers.status();
+                if (status.isInformational()) {
+                    if (!sawLeadingHeaders) {
+                        if (informationalHeadersBuilder == null) {
+                            informationalHeadersBuilder = ImmutableList.builder();
+                        }
+                        informationalHeadersBuilder.add(headers);
+                    }
+                    upstream.request(1);
+                } else {
+                    sawLeadingHeaders = true;
+                    completeInformationHeaders();
+                    completeHeaders(headers);
+                }
+                return;
+            }
+
+            if (httpObject instanceof HttpHeaders) {
+                final HttpHeaders trailers = (HttpHeaders) httpObject;
+                completeTrailers(trailers);
+                return;
+            }
+
+            assert downstream != null;
```

Comment:
could assign `downstream` to a local variable so that we remove IntelliJ warning